### PR TITLE
logic op dynamic state sample

### PR DESCRIPTION
**Pull request by Karina-Murawko-Wisniewska-Mobica**
## Description

This PR adds a sample for presenting possibilities to utilize feature included in the extension VK_EXT_extended_dynamic_state2: dynamic logic operations. It was agreed to create it as separate sample (the main sample for VK_EXT_extended_dynamic_state2 is added in PR https://github.com/KhronosGroup/Vulkan-Samples/pull/592).

Dynamic logic operations are applied to model on scene (blending effect s achieved) and logic operation can be changed using GUI (without need to create a new pipeline).

### Notes

Sample was tested using following configuration:
Windows 10 Pro
Nvidia T600 Laptop GPU (NVIDIA driver version: 31.0.15.2849, driver date: 02.02.2023)
Intel  i7-11850H
32GB RAM

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/extensions/logic_op_dynamic_state/README.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+/* Draw model */
+draw_created_model(draw_cmd_buffer);
+...
+
+VK_CHECK(vkEndCommandBuffer(draw_cmd_buffer));
+```
+
+## Enabling the required extensions
+
+The extended dynamic state 2 API requires Vulkan&reg; 1.0 and the appropriate headers / SDK is required. This extension has been [partially](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_extended_dynamic_state2.html#_promotion_to_vulkan_1_3) promoted to Vulkan&reg; 1.3.
```
Comment (SaschaWillems) [inline samples/extensions/logic_op_dynamic_state/README.md]: I don't think we need to use the ® in readmes. I'd remove it for better readability.

**Review by SaschaWillems — DISMISSED** [review]
Thank you very much for this sample. I did a review and this looks good to me. Runs fine on my setup (Win10, RTX 2060) and the readme is also looking fine.

I only have a very minor remark about the usage of the ® symbol in the readme. I think we should simply omit it.

File: samples/extensions/logic_op_dynamic_state/README.md (hunk 1 above)
Comment (Karina-Murawko-Wisniewska-Mobica) [reply samples/extensions/logic_op_dynamic_state/README.md]: Fixed, thanks.

**Comment by Karina-Murawko-Wisniewska-Mobica** [comment]
> Thank you very much for this sample. I did a review and this looks good to me. Runs fine on my setup (Win10, RTX 2060) and the readme is also looking fine.
> 
> I only have a very minor remark about the usage of the ® symbol in the readme. I think we should simply omit it.

Thanks for checking, I needed to push one more commit to fix validation layer errors (I had them enabled in a wrong way previously and missed the issues).

**Review by SaschaWillems — DISMISSED** [review]

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+	VkPipelineVertexInputStateCreateInfo vertex_input_state = vkb::initializers::pipeline_vertex_input_state_create_info();
+	vertex_input_state.vertexBindingDescriptionCount        = static_cast<uint32_t>(vertex_input_bindings.size());
+	vertex_input_state.pVertexBindingDescriptions           = vertex_input_bindings.data();
+	vertex_input_state.vertexAttributeDescriptionCount      = static_cast<uint32_t>(vertex_input_attributes.size());
+	vertex_input_state.pVertexAttributeDescriptions         = vertex_input_attributes.data();
+
+	std::array<VkPipelineShaderStageCreateInfo, 2> shader_stages{};
+	shader_stages[0] = load_shader("logic_op_dynamic_state/baseline.vert", VK_SHADER_STAGE_VERTEX_BIT);
+	shader_stages[1] = load_shader("logic_op_dynamic_state/baseline.frag", VK_SHADER_STAGE_FRAGMENT_BIT);
+C:
```
Comment (gpx1000) [inline samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp]: This label appears to be unused.  Is it perhaps a bug?

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+			for (int i = 0; i <= VK_LOGIC_OP_SET; ++i) /* VK_LOGIC_OP_SET is last operation in enum VkLogicOp */
+			{
+				names.push_back(vkb::to_string(static_cast<VkLogicOp>(i)));
+			}
+			return names;
+		}
+	} gui_settings;
+
+	LogicOpDynamicState();
+	~LogicOpDynamicState();
```
Comment (gpx1000) [inline samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h]: virtual class needs either virtual destructor or override.  Override is better in this case.

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	void update_uniform_buffers();
+	void create_pipeline();
+	void draw();
+	void load_assets();
+	void create_descriptor_pool();
+	void setup_descriptor_set_layout();
+	void create_descriptor_sets();
+	void model_data_creation();
+	void draw_created_model(VkCommandBuffer commandBuffer);
+	void on_update_ui_overlay(vkb::Drawer &drawer);
```
Comment (gpx1000) [inline samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h]: This is a virtual function, please use the override keyword here as well.

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+ * 	@fn void LogicOpDynamicState::create_render_context(vkb::Platform &platform)
+ * 	@brief Setting custom surface format priority list to required VK_FORMAT_B8G8R8A8_UNORM format
+ */
+void LogicOpDynamicState::create_render_context(vkb::Platform &platform)
+{
+	/* UNORM surface is required for logic operations */
+	auto surface_priority_list = std::vector<VkSurfaceFormatKHR>{
+	    {VK_FORMAT_B8G8R8A8_UNORM, VK_COLORSPACE_SRGB_NONLINEAR_KHR},
+	};
+	render_context = platform.create_render_context(*device.get(), surface, surface_priority_list);
```
Comment (gpx1000) [inline samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp]: get here is redundant on the smart pointer.

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+	                                          UINT32_MAX,
+	                                          4, 5, 7, 6,
+	                                          UINT32_MAX,
+	                                          2, 3, 6, 7};
+
+	struct
+	{
+		VkBuffer       buffer;
+		VkDeviceMemory memory;
+	} vertex_pos_staging, vertex_norm_staging, index_staging;
```
Comment (gpx1000) [inline samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp:617]: these are being used without being initialized.

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+
+/**
+ * @fn void LogicOpDynamicState::on_update_ui_overlay(vkb::Drawer &drawer)
+ * @brief Projecting GUI and transferring data between GUI and application
+ */
+void LogicOpDynamicState::on_update_ui_overlay(vkb::Drawer &drawer)
+{
+	if (drawer.header("Settings"))
+	{
+		if (drawer.combo_box("Logic operation", &gui_settings.selected_operation, gui_settings.logic_op_names))
```
Comment (gpx1000) [inline samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp]: logic_op_names is static.  It's better to access it like this: LogicOpDynamicState::GUI_settings::logic_op_names

**Review by gpx1000 — CHANGES_REQUESTED** [review]
This works for me in Linux.  Small changes are requested to eliminate a warning and some c++ issues.

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp (hunk 2 above)
Comment (Karina-Murawko-Wisniewska-Mobica) [reply samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp]: yes, a wrong window mistake, removed

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h (hunk 3 above)
Comment (Karina-Murawko-Wisniewska-Mobica) [reply samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h]: override added

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h (hunk 4 above)
Comment (Karina-Murawko-Wisniewska-Mobica) [reply samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.h]: override added

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp (hunk 5 above)
Comment (Karina-Murawko-Wisniewska-Mobica) [reply samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp]: changed to "get_device()"

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp (hunk 6 above)
Comment (Karina-Murawko-Wisniewska-Mobica) [reply samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp:617]: added initialization of members with VK_NULL_HANDLE

File: samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp (hunk 7 above)
Comment (Karina-Murawko-Wisniewska-Mobica) [reply samples/extensions/logic_op_dynamic_state/logic_op_dynamic_state.cpp]: fixed

**Comment by Karina-Murawko-Wisniewska-Mobica** [comment]
@gpx1000 
Comments addressed, thank you for review.

**Comment by gary-sweet** [comment]
Correctly reports not supported for me, but I can't actually test it.

**Review by gpx1000 — APPROVED** [review]

**Review by SaschaWillems — APPROVED** [review]

**Pull request merged**